File: girvan-newman/task1.py
from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession
from graphframes import GraphFrame
import os
import sys
import time


def main():
    os.environ["PYSPARK_SUBMIT_ARGS"] =  "--packages graphframes:graphframes:0.8.2-spark3.1-s_2.12 pyspark-shell"

    filter_threshold = int(sys.argv[1])
    input_file_path = sys.argv[2]
    output_file_path = sys.argv[3]

    conf = SparkConf().setAppName("task1")
    sc = SparkContext(conf=conf)
    spark = SparkSession(sc)
    spark.sparkContext.setLogLevel("ERROR")

    start_time = time.time()
    df = spark.read.csv(input_file_path, header=True, inferSchema=True)

    user_business_rdd = df.select("user_id", "business_id").rdd.map(lambda row: (row["user_id"], row["business_id"]))

    user_business_set_rdd = user_business_rdd.groupByKey().mapValues(set)

    user_pairs_rdd = user_business_set_rdd.cartesian(user_business_set_rdd)

    edges_rdd = (user_pairs_rdd.filter(lambda pair: pair[0][0] != pair[1][0]).filter(lambda pair: len(pair[0][1].intersection(pair[1][1])) >= filter_threshold).map(lambda pair: (pair[0][0], pair[1][0])).distinct())
    #edges = edges_rdd.collect()
    edges_df = edges_rdd.toDF(["src", "dst"])

    nodes_df = edges_df.select("src").union(edges_df.select("dst")).distinct().withColumnRenamed("src", "id")

    graph = GraphFrame(nodes_df, edges_df)


    communities = graph.labelPropagation(maxIter=5)

    sorted_communities = (communities.rdd.map(lambda row: (row["label"], row["id"])).groupByKey().mapValues(sorted).map(lambda x: x[1]).sortBy(lambda x: (len(x), x)).map(lambda x: [f"'{user}'" for user in x]).collect())

    with open(output_file_path, "w") as file:
        for community in sorted_communities:
            file.write(", ".join(community) + "\n")

    spark.stop()


if __name__ == "__main__":
    main()

File: girvan-newman/task1.scala
import org.apache.spark.sql.{SparkSession, Row}
import org.apache.spark.SparkConf
import org.graphframes.GraphFrame
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.functions._

object task1 {
  def main(args: Array[String]): Unit = {
    val filterThreshold = args(0).toInt
    val inputFilePath = args(1)
    val outputFilePath = args(2)

    // Configure Spark
    val conf = new SparkConf().setAppName("task1")
    val spark = SparkSession.builder().config(conf).getOrCreate()
    spark.sparkContext.setLogLevel("ERROR")

    val startTime = System.nanoTime()

    // Load data
    val df = spark.read.option("header", "true").csv(inputFilePath)
    val userBusinessRDD: RDD[(String, String)] = df.select("user_id", "business_id")
      .rdd.map(row => (row.getString(0), row.getString(1)))

    // Group businesses by user
    val userBusinessSetRDD = userBusinessRDD.groupByKey().mapValues(_.toSet)

    // Generate user pairs and filter based on the threshold
    val userPairsRDD = userBusinessSetRDD.cartesian(userBusinessSetRDD)
    val edgesRDD = userPairsRDD.filter { case ((user1, businesses1), (user2, businesses2)) =>
      user1 != user2 && businesses1.intersect(businesses2).size >= filterThreshold
    }.map { case ((user1, _), (user2, _)) => (user1, user2) }.distinct()

    // Convert edges to DataFrame
    import spark.implicits._
    val edgesDF = edgesRDD.toDF("src", "dst")

    // Create nodes DataFrame from edges
    val nodesDF = edgesDF.select("src").union(edgesDF.select("dst")).distinct().withColumnRenamed("src", "id")

    // Create GraphFrame
    val graph = GraphFrame(nodesDF, edgesDF)

    // Detect communities using Label Propagation Algorithm (LPA)
    val communities = graph.labelPropagation.maxIter(5).run()

    // Process communities and write to output file
    val sortedCommunities = communities.select("label", "id").rdd
      .map(row => (row.getLong(0), row.getString(1)))
      .groupByKey()
      .mapValues(users => users.toList.sorted)
      .sortBy { case (_, users) => (users.size, users) }
      .map { case (_, users) => users.map("'" + _ + "'").mkString(", ") }
      .collect()

    val file = new java.io.PrintWriter(outputFilePath)
    try {
      sortedCommunities.foreach(line => file.println(line))
    } finally {
      file.close()
    }

    val duration = (System.nanoTime() - startTime) / 1e9d
    println(f"Duration: $duration%.2f seconds")

    spark.stop()
  }
}
